File: Assignment3/domainJava/TranslatorAPIProlog.java
package domain;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.primefaces.model.TreeNode;

import domain.Attribute;
import domain.DomainData;
import domain.Entity;
import domain.Reference;
import domain.Relationship;

/**
 * A class that translates domain model data into Prolog facts.
 * This class is responsible for converting domain entities, relationships, and attributes
 * into Prolog syntax for knowledge representation.
 */
public class TranslatorAPIProlog implements Serializable {

    /** Serial version UID for serialization. */
    private static final long serialVersionUID = 1L;
    
    /** HashMap containing attributes for entities, indexed by entity name. */
    private HashMap<String, ArrayList<Attribute>> attributes;
    
    /** HashMap containing attributes for relationships, indexed by relationship name. */
    private HashMap<String, ArrayList<Attribute>> attributesRel;
    
    /** The name of the domain being translated. */
    private String domainName;
    
    /** Temporary string buffer used for accumulating recursive tree representation. */
    private String recur = "";

    /**
     * Constructor that creates a Prolog representation of a domain model.
     * 
     * @param domain The domain data model to be translated into Prolog facts
     * @throws IOException If an I/O error occurs during the process
     */
    public TranslatorAPIProlog(DomainData domain) throws IOException {
        String content = "domain(" + domain.getDomain().toLowerCase() + ").\n";
        content += createEntities(domain.getTopEntities());
        content += createRelationships(domain.getTopRelationships());
        
        // output originale
        
        content = writeFactsWithId(content, "id"); // Tartarello
        System.out.println(content);
        
        //TODO tartarello
        //OutputStream output = new FileOutputStream(output_path, false);
        //content = writeFactsWithId(content, "id");
        //output.write(content.getBytes());
        //output.close();
    }
    
    /**
     * Converts Prolog facts to facts with unique IDs.
     * Transforms each fact into the format fact(id_X, fact_content, 1).
     * 
     * @param content The string containing Prolog facts to be processed
     * @param id_prefix The prefix to use for the fact IDs
     * @return A string of Prolog facts with unique IDs
     */
    private static String writeFactsWithId(String content, String id_prefix) {
        int id = 0;
        String contentWithId = "";
        String[] facts = content.split("\n");
        for(String fact: facts) {
            fact = fact.substring(0, fact.length()-1);
            fact = "fact(" + id_prefix + "_" + id + ", " + fact + ", 1).\n";
            id++;
            contentWithId += fact;
        }
        return contentWithId;
    }

    /**
     * Creates Prolog facts for a list of entities and their attributes.
     * 
     * @param entities List of top-level entities to process
     * @return A string representation of entities and their attributes in Prolog syntax
     */
    private static String createEntities(List<Entity> entities) {
        String values = "";
        String att = "";
        for (Entity entity : entities) {
            values += "entity(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ").\n";
            for (Attribute attr : entity.getAttributes()) {
                att += "attribute(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                        + attr.getName().toLowerCase() + ", " + attr.getDataType().toLowerCase() + ").\n";
                if(!attr.getValues().isEmpty()) {
                    att += "values(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ", " + attr.getValuesToStringToLower() + ").\n";
                }
                if (attr.getMandatory() == true) {
                    att += "mandatory(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ").\n";
                }
                if (attr.isDisplay() == true) {
                    att += "display(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ").\n";
                }
                if (attr.isDistinguishing() == true) {
                    att += "distinguishing(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ").\n";
                }
                if (attr.getTarget() != null) {
                    att += "target(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", "
                            + attr.getName().toLowerCase() + ", " + attr.getTarget().toLowerCase() +").\n";
                }
            }
            for(Entity e : entity.getChildren()) {
                att += "parent(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                        + e.getName().toLowerCase() + ").\n";
                att += writeEntity(e);
            }
        }
        values += att;
        return values;
    }

    /**
     * Recursively processes an entity and its children, generating Prolog facts.
     * 
     * @param entity The entity to process
     * @return A string representation of the entity, its attributes, and children in Prolog syntax
     */
    private static String writeEntity(Entity entity) {
        String att = "";
        att += "entity(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ").\n";
        for (Attribute attr : entity.getNewAttributes()) {
            att += "attribute(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                    + attr.getName().toLowerCase() + ", " + attr.getDataType().toLowerCase() + ").\n";
            if(!attr.getValues().isEmpty()) {
                att += "values(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", "
                        + attr.getName().toLowerCase() + ", " + attr.getValuesToStringToLower() + ").\n";
            }        
            if (attr.getMandatory() == true) {
                att += "mandatory(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                        + attr.getName().toLowerCase() + ").\n";
            }
            if (attr.isDisplay() == true) {
                att += "display(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                        + attr.getName().toLowerCase() + ").\n";
            }
            if (attr.isDistinguishing() == true) {
                att += "distinguishing(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                        + attr.getName().toLowerCase() + ").\n";
            }
            if (attr.getTarget() != null) {
                att += "target(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", "
                        + attr.getName().toLowerCase() + ", " + attr.getTarget().toLowerCase() +").\n";
            }
        }
        for(Entity e : entity.getChildren()) {
            att += "parent(" + entity.getDomain().toLowerCase() + ", " + entity.getName().toLowerCase() + ", " 
                    + e.getName().toLowerCase() + ").\n";
            att += writeEntity(e);
        }
        return att;
    }

    /**
     * Creates Prolog facts for a list of relationships and their attributes.
     * 
     * @param relationships List of top-level relationships to process
     * @return A string representation of relationships and their attributes in Prolog syntax
     */
    private static String createRelationships(List<Relationship> relationships) {
        String att = "";
        for (Relationship relation : relationships) {
            for(Reference ref : relation.getReferences()) {
                att += "relationship(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() + ", " 
                        + ref.getSubject().toLowerCase() + ", " + ref.getObject().toLowerCase() + ").\n";
            }
            att += "inverse(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() + ", " 
                    + relation.getInverse().toLowerCase() + ").\n";
            for (Attribute attr : relation.getAttributes()) {
                att += "attribute(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() + ", " 
                        + attr.getName().toLowerCase() + ", " + attr.getDataType().toLowerCase() + ").\n";
                if(!attr.getValues().isEmpty()) {
                    att += "values(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ", " + attr.getValuesToStringToLower() + ").\n";
                }
                if (attr.getMandatory() == true) {
                    att += "mandatory(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ").\n";
                }
                if (attr.isDisplay() == true) {
                    att += "display(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ").\n";
                }
                if (attr.isDistinguishing() == true) {
                    att += "distinguishing(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() 
                                + ", " + attr.getName().toLowerCase() + ").\n";
                }
                if (attr.getTarget() != null) {
                    att += "target(" + relation.getDomain().toLowerCase() + ", " + relation.getName().toLowerCase() + ", " 
                            + attr.getName().toLowerCase() + ", " + attr.getTarget().toLowerCase() +").\n";
                }
            }
        }
        return att;
    }

    /**
     * Extracts class names from a tree node's children.
     * 
     * @param classi The tree node containing class information
     * @return An ArrayList of class names from the tree node's children
     */
    public ArrayList<String> getClassi(TreeNode classi) {
        ArrayList<String> s = new ArrayList<String>();
        List<TreeNode> children = classi.getChildren();
        for (TreeNode t : children) {
            s.add(t.toString());
        }
        return s;
    }

    /**
     * Gets the attributes map for entities.
     * 
     * @return HashMap mapping entity names to their attributes
     */
    public HashMap<String, ArrayList<Attribute>> getAttributes() {
        return attributes;
    }
    
    /**
     * Sets the attributes map for entities.
     * 
     * @param attributes HashMap mapping entity names to their attributes
     */
    public void setAttributes(HashMap<String, ArrayList<Attribute>> attributes) {
        this.attributes = attributes;
    }

    /**
     * Gets the attributes map for relationships.
     * 
     * @return HashMap mapping relationship names to their attributes
     */
    public HashMap<String, ArrayList<Attribute>> getAttributesRel() {
        return attributesRel;
    }
    
    /**
     * Sets the attributes map for relationships.
     * 
     * @param attributesRel HashMap mapping relationship names to their attributes
     */
    public void setAttributesRel(HashMap<String, ArrayList<Attribute>> attributesRel) {
        this.attributesRel = attributesRel;
    }

    /**
     * Gets the serial version UID.
     * 
     * @return The serial version UID value
     */
    public static long getSerialversionuid() {
        return serialVersionUID;
    }

    /**
     * Gets the domain name.
     * 
     * @return The name of the domain
     */
    public String getDomainName() {
        return domainName;
    }
    
    /**
     * Sets the domain name.
     * 
     * @param domainName The name to set for the domain
     */
    public void setDomainName(String domainName) {
        this.domainName = domainName;
    }

    /**
     * Recursively traverses a tree node and collects class names.
     * 
     * @param node The tree node to traverse
     * @return An ArrayList of class names found in the tree
     */
    public ArrayList<String> recursiveTree(TreeNode node) {
        ArrayList<String> classi = new ArrayList<String>();
        List<TreeNode> children = node.getChildren();
        for (int j = 0; j < node.getChildCount(); j++) {
            classi.add(children.get(j).toString());
            recursiveTree(children.get(j));
        }
        return classi;
    }

    /**
     * Recursively builds a taxonomy representation in Prolog format.
     * Results are accumulated in the 'recur' instance variable.
     * 
     * @param s The predicate name to use in the Prolog facts
     * @param node The current node in the tree
     * @param superV The parent/super class name
     */
    public void recursiveTreeTax(String s, TreeNode node, String superV) {
        List<TreeNode> children = node.getChildren();
        for (int j = 0; j < node.getChildCount(); j++) {
            recur += s + superV + ", " + children.get(j).toString() + ").\n";
            recursiveTreeTax(s, children.get(j), children.get(j).toString());
        }
    }

    /**
     * Gets the accumulated recursive representation.
     * 
     * @return The string containing the accumulated tree representation
     */
    public String getRecur() {
        return recur;
    }
    
    /**
     * Sets the accumulated recursive representation.
     * 
     * @param recur The string to set as the recursive representation
     */
    public void setRecur(String recur) {
        this.recur = recur;
    }

    /**
     * Converts an ArrayList of strings into a Prolog list representation.
     * 
     * @param lista The ArrayList of strings to convert
     * @return A string representing the ArrayList in Prolog list syntax
     */
    private String createList(ArrayList<String> lista) {
        String list = "";
        int i;
        list += "[";
        for (i =0; i < lista.size()-1; i++) {
            list += (lista.get(i)+", ");
        }
        list += lista.get(i);
        list += "]";
        return list;
    }

}

File: Assignment3/domain/TranslatorAPIProlog.py
import io
from typing import List, Dict, Optional, TYPE_CHECKING

# Assuming domain classes are in the same directory or package
from .Attribute import Attribute
from .DomainData import DomainData
from .Entity import Entity
from .Reference import Reference
from .Relationship import Relationship
if TYPE_CHECKING:
    from .TreeNode import TreeNode # Or from appropriate library

class TranslatorAPIProlog:

    serialVersionUID: int = 1  # Placeholder for Java's serialVersionUID
    attributes: Optional[Dict[str, List[Attribute]]] = None
    attributesRel: Optional[Dict[str, List[Attribute]]] = None
    domainName: Optional[str] = None
    recur: str = ""

    def __init__(self, domain: DomainData):
        """
        Initializes the TranslatorAPIProlog instance and generates Prolog facts.
        Mirrors the Java constructor.
        """
        if domain.getDomain() is None:
            raise ValueError("Domain name cannot be None")

        content = f"domain({domain.getDomain().lower()}).\n"
        content += TranslatorAPIProlog.createEntities(domain.getTopEntities())
        content += TranslatorAPIProlog.createRelationships(domain.getTopRelationships())

        # output originale
        content = TranslatorAPIProlog.writeFactsWithId(content, "id")
        print(content)


    @staticmethod
    def writeFactsWithId(content: str, id_prefix: str) -> str:
        """
        Adds a unique ID to each fact string.
        Mirrors the Java static method.
        """
        id_counter = 0
        contentWithId = ""
        # Split carefully, handling potential empty lines
        facts = [line for line in content.splitlines() if line.strip()]
        for fact in facts:
            fact = fact[:-1]
            # Format with ID
            fact_with_id = f"fact({id_prefix}_{id_counter}, {fact}, 1).\n"
            id_counter += 1
            contentWithId += fact_with_id
        return contentWithId

    @staticmethod
    def createEntities(entities: List[Entity]) -> str:
        """
        Generates Prolog facts for top-level entities and their attributes recursively.
        Mirrors the Java static method.
        """
        values = ""
        att = ""
        for entity in entities:
            domain_lower = entity.getDomain().lower() if entity.getDomain() else "unknown_domain"
            name_lower = entity.getName().lower() if entity.getName() else "unknown_entity"

            values += f"entity({domain_lower}, {name_lower}).\n"
            # Use getAttributes() for top-level entities as in Java
            for attr in entity.getAttributes():
                attr_name_lower = attr.getName().lower() if attr.getName() else "unknown_attr"
                attr_dtype_lower = attr.getDataType().lower() if attr.getDataType() else "unknown_type"

                att += f"attribute({domain_lower}, {name_lower}, {attr_name_lower}, {attr_dtype_lower}).\n"
                if attr.getValues(): # Check if list is not empty
                    # Ensure getValuesToStringToLower exists and returns a Prolog-compatible list string
                    values_str = attr.getValuesToStringToLower() # Assuming this method exists and formats correctly
                    if values_str: # Check if the result is not empty
                         att += f"values({domain_lower}, {name_lower}, {attr_name_lower}, {values_str}).\n"

                if attr.getMandatory():
                    att += f"mandatory({domain_lower}, {name_lower}, {attr_name_lower}).\n"
                if attr.isDisplay():
                    att += f"display({domain_lower}, {name_lower}, {attr_name_lower}).\n"
                if attr.isDistinguishing():
                    att += f"distinguishing({domain_lower}, {name_lower}, {attr_name_lower}).\n"
                if attr.getTarget():
                    target_lower = attr.getTarget().lower() # type: ignore
                    att += f"target({domain_lower}, {name_lower}, {attr_name_lower}, {target_lower}).\n"

            # Process children recursively
            children = entity.getChildren()
            if children:
                 for e in children:
                      child_name_lower = e.getName().lower() if e.getName() else "unknown_child"
                      att += f"parent({domain_lower}, {name_lower}, {child_name_lower}).\n"
                      # Call writeEntity for children as in Java
                      att += TranslatorAPIProlog.writeEntity(e)
        values += att
        return values

    @staticmethod
    def writeEntity(entity: Entity) -> str:
        """
        Generates Prolog facts for a specific entity and its *new* attributes recursively.
        Mirrors the Java static method.
        """
        att = ""
        domain_lower = entity.getDomain().lower() if entity.getDomain() else "unknown_domain"
        name_lower = entity.getName().lower() if entity.getName() else "unknown_entity"

        att += f"entity({domain_lower}, {name_lower}).\n"
        # Use getNewAttributes() for child entities as in Java
        for attr in entity.getNewAttributes():
            attr_name_lower = attr.getName().lower() if attr.getName() else "unknown_attr"
            attr_dtype_lower = attr.getDataType().lower() if attr.getDataType() else "unknown_type"

            att += f"attribute({domain_lower}, {name_lower}, {attr_name_lower}, {attr_dtype_lower}).\n"
            if attr.getValues():
                values_str = attr.getValuesToStringToLower() # Assuming this method exists
                if values_str:
                    att += f"values({domain_lower}, {name_lower}, {attr_name_lower}, {values_str}).\n"
            if attr.getMandatory():
                att += f"mandatory({domain_lower}, {name_lower}, {attr_name_lower}).\n"
            if attr.isDisplay():
                att += f"display({domain_lower}, {name_lower}, {attr_name_lower}).\n"
            if attr.isDistinguishing():
                att += f"distinguishing({domain_lower}, {name_lower}, {attr_name_lower}).\n"
            if attr.getTarget():
                target_lower = attr.getTarget().lower() # type: ignore
                att += f"target({domain_lower}, {name_lower}, {attr_name_lower}, {target_lower}).\n"

        # Process children recursively
        children = entity.getChildren()
        if children:
            for e in children:
                child_name_lower = e.getName().lower() if e.getName() else "unknown_child"
                att += f"parent({domain_lower}, {name_lower}, {child_name_lower}).\n"
                att += TranslatorAPIProlog.writeEntity(e)
        return att

    @staticmethod
    def createRelationships(relationships: List[Relationship]) -> str:
        """
        Generates Prolog facts for relationships, inverses, and attributes.
        Mirrors the Java static method.
        """
        att = ""
        for relation in relationships:
            domain_lower = relation.getDomain().lower() if relation.getDomain() else "unknown_domain"
            rel_name_lower = relation.getName().lower() if relation.getName() else "unknown_relation"

            # Relationship facts from references
            refs = relation.getReferences()
            if refs:
                for ref in refs:
                    subj_lower = ref.getSubject().lower() if ref.getSubject() else "unknown_subject"
                    obj_lower = ref.getObject().lower() if ref.getObject() else "unknown_object"
                    att += f"relationship({domain_lower}, {rel_name_lower}, {subj_lower}, {obj_lower}).\n"

            # Inverse relationship fact
            inv_lower = relation.getInverse().lower() if relation.getInverse() else "unknown_inverse"
            att += f"inverse({domain_lower}, {rel_name_lower}, {inv_lower}).\n"

            # Attributes of the relationship
            rel_attrs = relation.getAttributes()
            if rel_attrs:
                for attr in rel_attrs:
                    attr_name_lower = attr.getName().lower() if attr.getName() else "unknown_attr"
                    attr_dtype_lower = attr.getDataType().lower() if attr.getDataType() else "unknown_type"

                    att += f"attribute({domain_lower}, {rel_name_lower}, {attr_name_lower}, {attr_dtype_lower}).\n"
                    if attr.getValues():
                        values_str = attr.getValuesToStringToLower() # Assuming this method exists
                        if values_str:
                             att += f"values({domain_lower}, {rel_name_lower}, {attr_name_lower}, {values_str}).\n"
                    if attr.getMandatory():
                        att += f"mandatory({domain_lower}, {rel_name_lower}, {attr_name_lower}).\n"
                    if attr.isDisplay():
                        att += f"display({domain_lower}, {rel_name_lower}, {attr_name_lower}).\n"
                    if attr.isDistinguishing():
                        att += f"distinguishing({domain_lower}, {rel_name_lower}, {attr_name_lower}).\n"
                    if attr.getTarget():
                        target_lower = attr.getTarget().lower() # type: ignore
                        att += f"target({domain_lower}, {rel_name_lower}, {attr_name_lower}, {target_lower}).\n"
        return att

    # --- Instance Methods ---

    def getClassi(self, classi: TreeNode) -> List[str]:
        """
        Gets the string representation of the direct children of a TreeNode.
        Requires a TreeNode implementation with getChildren() and __str__.
        """
        s: List[str] = []
        children = classi.getChildren() # Assuming getChildren exists
        if children:
            for t in children:
                s.append(str(t)) # Assuming __str__ gives the desired representation
        return s

    def getAttributes(self) -> Optional[Dict[str, List[Attribute]]]:
        return self.attributes

    def setAttributes(self, attributes: Dict[str, List[Attribute]]) -> None:
        self.attributes = attributes

    def getAttributesRel(self) -> Optional[Dict[str, List[Attribute]]]:
        return self.attributesRel

    def setAttributesRel(self, attributesRel: Dict[str, List[Attribute]]) -> None:
        self.attributesRel = attributesRel

    # getSerialversionuid is Java specific, omitted.

    def getDomainName(self) -> Optional[str]:
        return self.domainName

    def setDomainName(self, domainName: str) -> None:
        self.domainName = domainName

    def recursiveTree(self, node: 'TreeNode') -> List[str]:
        """
        Recursively collects the string representation of all nodes in a subtree.
        Requires a TreeNode implementation.
        """
        classi: List[str] = []
        children = node.getChildren() # Assuming getChildren exists
        if children:
            # Java code iterates using getChildCount, Python can iterate directly
            for child_node in children:
                classi.append(str(child_node)) # Add current child
                classi.extend(self.recursiveTree(child_node)) # Recurse and add results
        return classi

    def recursiveTreeTax(self, s: str, node: 'TreeNode', superV: str) -> None:
        """
        Recursively builds a string representation of a taxonomy.
        Modifies the instance variable `recur`. Requires a TreeNode implementation.
        """
        children = node.getChildren() # Assuming getChildren exists
        if children:
            # Java code iterates using getChildCount, Python can iterate directly
            for child_node in children:
                child_str = str(child_node)
                self.recur += f"{s}{superV}, {child_str}).\n"
                self.recursiveTreeTax(s, child_node, child_str)

    def getRecur(self) -> str:
        return self.recur

    def setRecur(self, recur: str) -> None:
        self.recur = recur

    # Private helper method in Java, made internal (_ prefix) in Python
    def _createList(self, lista: List[str]) -> str:
        """
        Creates a Prolog-style list string from a Python list of strings.
        Mirrors the Java private method createList.
        """
        if not lista:
            return "[]"
        list_content = ", ".join(lista)
        return f"[{list_content}]"
